### PR TITLE
[Improvement] Fixes #784 and use random generator objects rather than np.random global RNG

### LINKED ISSUE #784: nk.eda_simulate non reproducible even when random_state is set to a constant
**Describe the bug**
`nk.eda_simulate` generates random signals even when `random_state` is set to a constant

**To Reproduce**
```python
>>> eda_signal = nk.eda_simulate(duration=30, scr_number=5, drift=0.1, random_state=1234)
>>> print(eda_signal)
[0.98571118 0.98526776 0.98387649 ... 3.9812714  3.9785683  3.97764784]
>>> eda_signal = nk.eda_simulate(duration=30, scr_number=5, drift=0.1, random_state=1234)
>>> print(eda_signal)
[0.99979647 0.99998727 1.00034615 ... 4.05665343 4.05670009 4.05676347]
```

**Expected behaviour**
Setting `random_state` should make `nk.eda_simulate`  always return the same signal.
The problem is that `nk.eda_simulate` calls `signal_distort` (line [89](https://github.com/neuropsychology/NeuroKit/blob/8c564938c5fc2b4a377fd89bb14ac07a430983b9/neurokit2/eda/eda_simulate.py#L89)) without setting its `random_state`. This means that the argument is set to its default `None` and therefore line [106](https://github.com/neuropsychology/NeuroKit/blob/8c564938c5fc2b4a377fd89bb14ac07a430983b9/neurokit2/signal/signal_distort.py#L106) resets the random generator to a random state.
A solution is to call `signal_distort` with 'random_state=np.random.randint(np.iinfo(np.uint32).max)`.
An even better solution, moving forward, would be to use a see generator object and pass it around from one function to all its callees, as suggested for example [here](https://albertcthomas.github.io/good-practices-random-number-generators/).
A simple way to do this is to replace the
```python
np.random.seed(random_state)
```
at the beginning of all functions and replace it with:
```python
rng = np.random.default_rng(random_state)
```
and use `rng` throughout the function (e.g. we need to replace `np.random.normal(0, 5, size=1)` with `rng.normal(0, 5, size=1)`) and pass it to all callees as `random_state=rng`. We also need to remove `np.random.seed(None)` from the end of each function (since we are not touching the singleton numpy random generator, there's no need to reset it to a random state).

### PR DESCRIPTION
**Pull request by lciti**
# Description

This PR fixes #784 and make changes to the way random numbers are generated along the lines of what I describe in the discussion of #784.

# Proposed Changes

Modified the approach to random number generation (also solving the issue that the same seed was used both to generate a signal and its noise).
XXX_simulate functions now use an RNG object (RandomState or Generator) rather than the global singleton numpy.random one.
Their argument` random_state` can take several values, notably the following ones:
- When this is an integer, a RandomState object is created and used (less performant but more reproducible).
- When this is None, a Generator object is created and used (more performant but without reproducibility guarantees for different versions of numpy, it is also the recommended RNG for new code by the [numpy manual](https://numpy.org/doc/stable/reference/random/index.html)).
- When it's already a RandomState or Generator, this object is used to generate random values.

Some ways to generate random numbers had to be changed to use methods available both in RandomState and in Generator, e.g. `randn`->`standard_normal` (since `randn` is in RandomState but not in Generator, while `standard_normal` is in both).

Regarding the issue that the same seed was used both to generate a signal and its noise, now the XXX_simulate functions also take `random_state_distort`, which can take value "legacy" (to reproduce results up to nk 0.2.3) or "spawn" to generate noise independent of the signal. It can also take, e.g., a seed, which allows creating different simulated signals with the same "signal" and different random instances of "noise" added to it. This may be useful to determine whether a given processing algorithm (e.g. peak detection) produces similar results when fed the same signal corrupted by different random instances of noise.

*NOTE:* I haven't added any documentation or example of the extra functionality or extra argument yet. Before investing additional time, I want to make sure @DominiqueMakowski is happy with the changes.

# Checklist

Here are some things to check before creating the PR. If you encounter any issues, do let us know :)

- [x] I have read the [CONTRIBUTING](https://github.com/neuropsychology/NeuroKit/blob/master/.github/CONTRIBUTING.rst#structure-and-code) file.
- [x] My PR is targetted at the **dev branch** (and not towards the master branch).
- [x] I ran the [CODE CHECKS](https://github.com/neuropsychology/NeuroKit/blob/master/.github/CONTRIBUTING.rst#run-code-checks) on the files I added or modified and fixed the errors.
- [ ] I have added the newly added features to **News.rst** (if applicable)

### CODE REVIEW

**Comment by lciti** [comment]
Failed the new test for k-means. It looks like the test suite uses a version of sk-learn that does not implement the [`n_init='auto'` argument](https://github.com/scikit-learn/scikit-learn/commit/0d923daf6ddbf1afc0915a145955cfff97505f51), which `nk` assumes available here: https://github.com/neuropsychology/NeuroKit/blob/8c564938c5fc2b4a377fd89bb14ac07a430983b9/neurokit2/stats/cluster.py#L190
It works fine on my machine with sk-learn 1.2.1.

File: tests/tests_rsp.py [hunk 1]
```diff
@@ -31,6 +32,60 @@ def test_rsp_simulate():
...
+        (13579, "spawn"),
+        (13579, 24680),
+        (13579, None),
+        (np.random.RandomState(33), "spawn"),
+        (np.random.SeedSequence(33), "spawn"),
+        (np.random.Generator(np.random.Philox(33)), "spawn"),
+        (None, "spawn"),
+    ],
+)
+def test_ppg_simulate_all_rng_types(random_state, random_state_distort):
```
Comment (danibene) [inline tests/tests_rsp.py:70]: I think you meant rsp, not ppg

**Comment by danibene** [comment]
> Failed the new test for k-means. It looks like the test suite uses a version of sk-learn that does not implement the [`n_init='auto'` argument](https://github.com/scikit-learn/scikit-learn/commit/0d923daf6ddbf1afc0915a145955cfff97505f51), which `nk` assumes available here: https://github.com/neuropsychology/NeuroKit/blob/8c564938c5fc2b4a377fd89bb14ac07a430983b9/neurokit2/stats/cluster.py#L190
> It works fine on my machine with sk-learn 1.2.1.

@DominiqueMakowski how would you suggest addressing this? Changing the version of sklearn used in the test suite/required by neurokit, or changing the code e.g. so that the number of clusters is provided as an integer if `n_init=auto` fails?

**Review by danibene — COMMENTED** [review]
This generally looks good to me, I think all that remains to do is renaming the rsp test function from ppg, adding documentation, and addressing the failing cluster test

File: tests/tests_rsp.py (hunk 1 above)
Comment (lciti) [reply tests/tests_rsp.py:70]: Yes, I think you are right.

File: neurokit2/misc/random.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    # If seed is already a random number generator return it as it is
+    if isinstance(seed, (np.random.Generator, np.random.RandomState)):
+        return seed
+    # If seed is something else, use the new Generator class
+    # Note: to initialise the new generator class with an integer seed, use, e.g.:
+    # check_rng(np.random.SeedSequence(123))
+    return np.random.default_rng(seed)
+
+
+def spawn_rng(rng, n_children=1):
```
Comment (DominiqueMakowski) [inline neurokit2/misc/random.py:22]: This could probably be made an internal `_spawn_rng` and not exported as it's only used in get_children_rng afaik

**Comment by DominiqueMakowski** [comment]
Thanks for that impressive PR! 

Maybe we could change the names, what do you think about `sanitize_randomseed()` (for check_rng) and `multiple_randomseeds()`?

File: neurokit2/misc/random.py (hunk 2 above)
Comment (lciti) [reply neurokit2/misc/random.py:22]: It is only used internally for now but I would leave it as public as spawning an rng is good practice whenever several RNGs are needed. For example, if one needs to simulate several signals (e.g. several patients in a study) one should not reuse the same seed or use seed, seed+1, seed+2... but rather 
```python
master_seed = 42
patients_RNGs = nk.misc.spawn_rng(master_seed, n_children=n_patients)
PPGs = []
for i in range(n_patients):
    PPGs.append(nk.ppg_simulate(..., random_state=patients_RNGs[i]))
```

**Comment by lciti** [comment]
@DominiqueMakowski I agree the names are a bit strange. However I called them this way based on the names of virtually identical functions in other python projects. See for example [sklearn's `check_random_state`](https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_random_state.html), [statsmodels's one](https://www.statsmodels.org/dev/generated/statsmodels.tools.rng_qrng.check_random_state.html) and [scipy's one](https://github.com/scipy/scipy/blob/0655c14c549f71200ed6a8a4938098270c4352dc/scipy/_lib/_util.py#L177), which do virtually the same thing. I decided to call it `check_rng` rather than `check_random_state` to make it closer to the new recommended numpy Generator API, however we could also call it `check_random_state` for consistency with these other projects. To be honest, I initially called it `get_rng` (or `get_random_state`) but then I found the sklearn's one and tried to be consistent. Anyway, I am open to reconsider the choice (as I am not 100% sure myself).
As for `spawn`, that's how the new numpy API will call this function, see [here](https://github.com/numpy/numpy/pull/23195). The function `spawn_rng` I implemented basically calls the numpy `spawn` method when available (only in the next release and only for `Generator`, not for `RandomState`) and does the same thing or something close in the other cases. In this case I think `spawn` is the right term (but could be `spawn_random_state`).

**Comment by lciti** [comment]
@danibene , @DominiqueMakowski
It's ready to go on my side. Please let me know if you have any comments.

**Comment by danibene** [comment]
This looks good to me!

**Comment by DominiqueMakowski** [comment]
Thanks a lot @lciti for that impressive work :) merging now

**Comment by lciti** [comment]
My pleasure! Many thanks to you for developing and maintaining such an impressive tool!
I hope we can fix that other bug too (https://github.com/neuropsychology/NeuroKit/issues/781)!
